### ISSUE TITLE
New analyzer reports error "Unexpected keyword argument" in dataclass

### ISSUE THREAD

**Issue by miki725** [python/mypy#7250]
After upgrading mypy to latest `0.720` where new analyzer is the default, it seems to break my usage of a `dataclass`. Cannot replicate my issue in a small code example so not sure what is the exact root cause so will simply link to my source code. Perhaps im using mypy incorrectly or maybe its some side-effect in new analyzer.

The differences:

```
➜ python --version
Python 3.7.3

➜ mypy --version
mypy 0.720

➜ mypy importanize --no-new-semantic-analyzer

➜ mypy importanize --new-semantic-analyzer
...
importanize/config.py:97: error: Unexpected keyword argument "formatter" for "Config"
importanize/config.py:152: error: Unexpected keyword argument "formatter" for "Config"
...
```

This report can focus on the differences in `config.py`. Relevant source code lines: [config.py:97](https://github.com/miki725/importanize/blob/30cee5e465ecf994c61787461cdaed37de5d7b08/importanize/config.py#L160-L162) and [config.py:152](https://github.com/miki725/importanize/blob/30cee5e465ecf994c61787461cdaed37de5d7b08/importanize/config.py#L105-L109). Quick tldr of the [`Config`](https://github.com/miki725/importanize/blob/30cee5e465ecf994c61787461cdaed37de5d7b08/importanize/config.py#L67-L79)

```python
class Config:
    path: typing.Union[pathlib.Path, None] = None
    after_imports_new_lines: int = 2
    length: int = 80
    formatter: typing.Type[formatters.Formatter] = formatters.GroupedFormatter
    groups: typing.Iterable[GroupConfig] = (
        GroupConfig(type="stdlib"),
        GroupConfig(type="sitepackages"),
        GroupConfig(type="remainder"),
        GroupConfig(type="local"),
    )
    exclude: typing.Iterable[str] = ("*/.tox/*",)
    add_imports: typing.Iterable[ImportStatement] = ()
```

`formatter` is a valid argument and code runs perfectly fine. Not sure why mypy only complains on this argument and others are fine.

Difference is still observed with mypy installed from git master:

```
➜ python3 -m pip install -U git+git://github.com/python/mypy.git
...
Successfully installed mypy-0.730+dev.0a7cb9422d105c7099fdf4ef6a84765ffdcf3058

➜ mypy importanize --new-semantic-analyzer
...
importanize/config.py:97: error: Unexpected keyword argument "formatter" for "Config"
importanize/config.py:152: error: Unexpected keyword argument "formatter" for "Config"
...
```

Here is my current [mypy config](https://github.com/miki725/importanize/blob/30cee5e465ecf994c61787461cdaed37de5d7b08/setup.cfg#L23-L45):

```
[mypy]
python_version=3.7
new_semantic_analyzer=False
# strict
strict_optional=False
no_implicit_optional=False
strict_equality=True
implicit_reexport=False
allow_redefinition=False
# warnings
warn_unused_configs=True
warn_redundant_casts=True
warn_unused_ignores=True
warn_return_any=True
# any
disallow_subclassing_any=True
disallow_any_generics=True
# untyped
disallow_untyped_calls=True
disallow_untyped_defs=True
disallow_incomplete_defs=True
check_untyped_defs=True
disallow_untyped_decorators=True
```

Let me know if I can provide any additional information.

Thanks

**Comment by ilevkivskyi** [comment]
Could you please provide a simple self-contained repro for this?

**Comment by miki725** [comment]
sorry but could not reproduce the issue in self-contained example. it seems to work as expected in small examples I tried but in my library it consistently does not work. not sure how I can replicate issue in small example.

**Comment by ilevkivskyi** [comment]
OK, then my guess it is something to do with import cycles.

**Comment by olmokramer** [comment]
I've come across the same issue. Here's a minimal reproducer:

```py
import typing as t
import dataclasses


@dataclasses.dataclass
class BaseClass:
    x: t.Optional['SubClass']


@dataclasses.dataclass
class SubClass(BaseClass):
    y: str


a = SubClass(None)
```

Python version: 3.7.4
Mypy version: 0.720

**Comment by ilevkivskyi** [comment]
@olmokramer The first call in your example actually fails at runtime but the second one is indeed a simple repro for the issue. I know understand the cause and it looks pretty bad. The fix however may be very simple, if so, I will submit a PR shortly.

**Comment by olmokramer** [comment]
@ilevkivskyi Whoops, forgot to check that it actually runs. Great that you could find the cause and implement a fix so fast. Thanks!
